File: trivia.py
# Python Version: Played in the terminal

import requests
import os
from html import unescape
from random import shuffle

# response = requests.get("https://opentdb.com/api.php?amount=1&category=")

categories = requests.get("https://opentdb.com/api_category.php").json()["trivia_categories"]

def print_categories(categories):
    print(15 * "-" + "CATEGORIES" + 15 * "-")
    for i, obj in enumerate(categories):
        print(f"{i + 1}: {obj['name']}")
        
    return i + 1
    
def get_question(cat):
    response = requests.get("https://opentdb.com/api.php?amount=1&category=" + str(cat))
    question_dict = response.json()['results'][0]
    if (question_dict['type'] == 'boolean'):
        print("True or False: ", end="")
    print(unescape(question_dict['question']))
    if (question_dict['type'] == 'multiple'):
        choices: list = question_dict['incorrect_answers'].copy()
        choices.append(question_dict['correct_answer'])
        shuffle(choices)
        for let, choice in enumerate(choices):
            print(f'{chr(let + 97)}) {unescape(choice)}')
        question_dict.update({"choices" : choices})
    return question_dict

def check_answer(answer, question_dict):
    answer = answer.lower()
    if (question_dict['type'] == 'boolean'):
        return answer == question_dict['correct_answer'][0].lower()
    else:
        return question_dict['correct_answer'] == question_dict['choices'][ord(answer) - 97]


while True:
    maximum = print_categories(categories)
    while True:
        try:
            cat = int(input("Please type in the number corresponding to the category: "))
            if (cat < 1) or (cat > maximum):
                raise Exception
            break
        except:
            print("Invalid category")
        
    os.system('cls')
    question_dict = get_question(cat + 8)
    while True:
        answer = input("Answer: ")
        if len(answer) != 0:
            if question_dict['type'] == 'boolean' and (answer[0].lower() == 't' or answer[0].lower() == 'f'):
                break
            elif question_dict['type'] == 'multiple' and (97 <= ord(answer[0].lower()) <= 100):
                break
        print("Invalid answer")
    if check_answer(answer[0], question_dict):
        print("Correct!")
    else:
        print("Incorrect")
        print(f"Correct answer was: {question_dict['correct_answer']}")
    
    answer = input("Would you like to play again (y/n)?: ").lower()
    if (len(answer) == 0) or (answer[0] != 'y'):
        break
    
        

        

File: trivia.js
class Question {
    constructor(question) {
        this.type = question.type;
        this.question = question.question;
        this.correct_answer = question.correct_answer;
        this.answers = [];
        if (this.type == "multiple") {
            for (let answer of question.incorrect_answers) {
                this.answers.push(answer);
            }
            this.answers.push(question.correct_answer);
            shuffle(this.answers);
        } else {
            this.answers.push("True", "False");
        }
    }

    display () {
        document.getElementById("grade").innerHTML = "";

        let text = "";
        if (this.type == "boolean") {
            text += "True or False: ";
        }
        text += this.question;
        document.getElementById("question").innerHTML = text;

        let html_answers = "";
        for (let answer of this.answers) {
            html_answers += `<div><input type="radio" id="${answer}" value="${answer}" name="answer"><label for="${answer}">${answer}</label></div>`;
        }
        html_answers += `<div><input id="submit-answer" type="submit" value="Submit"></div>`;
        document.getElementById("answers").innerHTML = html_answers;
        document.getElementById("answers").addEventListener("submit", e => {
            let form_data = new FormData(e.target);
            let user_answer = form_data.get("answer");
            if (html_decode(this.correct_answer) == user_answer) {
                document.getElementById("grade").innerHTML = "Correct!";
                score += 1;
                document.getElementById("score").innerHTML = `Score = ${score}`;
            } else {
                document.getElementById("grade").innerHTML = "Incorrect";
            }
            document.getElementById("submit-answer").style.display = "none";
            e.preventDefault();
        }, {once: true});
    }
}

function shuffle(array) {
    let currentIndex = array.length;
  
    // While there remain elements to shuffle...
    while (currentIndex != 0) {
  
      // Pick a remaining element...
      let randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex--;
  
      // And swap it with the current element.
      [array[currentIndex], array[randomIndex]] = [
        array[randomIndex], array[currentIndex]];
    }
}

function html_decode(st) {
    let doc = new DOMParser().parseFromString(st, "text/html");
    return doc.documentElement.textContent;
}

async function make_request(url) {
    let obj;

    const res = await fetch(url);

    obj = res.json();
    return obj;
}

const base_url = "https://opentdb.com/api.php?amount=1&category=";
let score = 0;

make_request("https://opentdb.com/api_category.php")
    .then(data => {
        let val = 9;
        for (let category of data.trivia_categories) {
            document.getElementById("category").innerHTML += `<option value="${val}">` + category.name + "</option>";
            val++;
        }
        
    });

document.getElementById("form").addEventListener("submit", (e) => {
    let form_data = new FormData(e.target);
    let cat = form_data.get("category");
    make_request(base_url + cat)
        .then(data => {
            let question = new Question(data.results[0]);
            question.display();
        })
    e.preventDefault();
});